hpx/lcos/broadcast_async.hpp

```diff
@@ -0,0 +1,211 @@
+                    hpx::detail::name_from_basename(name, site)) ==
+                hpx::get_locality_id()
+            );
+
+            // find_from_basename is always a local operation (see assert above)
+            typedef typename std::decay<T>::type result_type;
+            return hpx::dataflow(
+                [](hpx::future<hpx::id_type> f, result_type && t) -> void
+                {
+                    set_lco_value(f.get(), std::move(t));
+                },
+                hpx::find_from_basename(name, site), std::forward<T>(t));
+        }
+
+        template <typename T>
+        struct broadcast_there
+        {
+            static hpx::future<void> call(std::string const& name,
+                std::vector<std::size_t> const& sites, T && t)
+            {
+                // apply actual broadcast operation to set of sites managed
+                // on this locality
+                if (sites.size() == 1)
+                {
+                    return detail::broadcast_there_invoke(
+                        name, sites[0], std::move(t));
+                }
+
+                std::vector<hpx::future<void> > futures;
+                futures.reserve(sites.size());
+                for(std::size_t i : sites)
+                {
+                    futures.push_back(detail::broadcast_there_invoke(name, i, t));
+                }
+                return hpx::when_all(futures);
+            }
+        };
+
+        template <typename T>
+        struct make_broadcast_there_action
+        {
+            typedef typename HPX_MAKE_ACTION(broadcast_there<T>::call)::type type;
+        };
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    hpx::future<void> broadcast_send(char const* basename, T && t,
+        std::size_t num_sites = std::size_t(-1),
+        std::size_t generation = std::size_t(-1))
+    {
+        if (num_sites == std::size_t(-1))
+            num_sites = hpx::get_num_localities(hpx::launch::sync);
+
+        std::string name(basename);
+        if (generation != std::size_t(-1))
+            name += std::to_string(generation) + "/";
+
+        // generate mapping of which sites are managed by what symbol namespace
+        // instances
+        std::map<std::uint32_t, std::vector<std::size_t> > locality_indices =
+            detail::generate_locality_indices(name, num_sites);
+
+        typedef typename lcos::detail::make_broadcast_there_action<
+                typename std::decay<T>::type
+            >::type broadcast_there_action;
+
+        if (locality_indices.size() == 1)
+        {
+            return hpx::async(broadcast_there_action(),
+                hpx::naming::get_id_from_locality_id(
+                    locality_indices.begin()->first
+                ), std::move(name), std::move(locality_indices.begin()->second),
+                std::forward<T>(t));
+        }
+
+        std::vector<hpx::future<void> > futures;
```

Comment:
Unfortunately that does not quite fit the bill. It would amount to sending the `std::map<std::uint32_t, std::vector<std::size_t> >` to everybody which is much more than needed... I was contemplating about creating a special broadcast which splits off the map, but wanted to leave that to future optimizations.